protocol/contracts/leaser/src/leaser.rs

```diff
@@ -121,28 +121,36 @@ where
     })
 }
 
-pub(super) fn try_close_protocol<ProtocolsRegistryLoader>(
-    storage: &dyn Storage,
+pub(super) fn try_close_protocol<ProtocolsRegistryLoader, MsgFactory>(
+    storage: &mut dyn Storage,
+    new_lease_code: Code,
+    max_leases: MaxLeases,
+    migrate_msg: MsgFactory,
     protocols_registry: ProtocolsRegistryLoader,
     migration_spec: ProtocolContracts<MigrationSpec>,
     force: ForceClose,
 ) -> ContractResult<MessageResponse>
 where
+    MsgFactory: Fn(Addr) -> MigrateMsg,
     ProtocolsRegistryLoader: FnOnce(&dyn Storage) -> ContractResult<Addr>,
 {
     if force == ForceClose::No && has_lease(storage) {
```

Comment:
[Secure coding] Prefer matching against the safer case implicitly and explicitly against the more dangerous ones. That way when a new variant is added, if it happens for us to forget to update it, it will still go the safe route.  
Another way to avoid it is to use exhaustive pattern matching, but it might not be the perfect solution for this case.

In this case, the safer thing here is to ask "is it forced? if not, are there any more leases? if not, close".  
One way might be the following suggestion and inverting the `if` and `else` bodies:
```suggestion
    if matches!(force, ForceClose::KillProtocol) || !has_lease(storage) {
```

Reply:
Re: "In this case, the safer thing here is to ask "is it forced? if not, are there any more leases? if not, close"." This is what I do.
The rationale here is to use easier-to-read code and avoid negative checks.

Reply:
in addition, that is covered with tests that validate the forcing logic

Reply:
refactored it in another way to address what you say and to avoid lease migration when force=no